d2graph/d2graph.go

```diff
@@ -1224,28 +1219,49 @@ func (g *Graph) SetDimensions(mtexts []*d2target.MText, ruler *textmeasure.Ruler
 			return err
 		}
 
-		obj.Width = float64(go2.Max(defaultDims.Width, desiredWidth))
-		obj.Height = float64(go2.Max(defaultDims.Height, desiredHeight))
-
-		paddingX, paddingY := obj.GetPadding()
+		shapeType := d2target.DSL_SHAPE_TO_SHAPE_TYPE[dslShape]
+		contentBox := geo.NewBox(geo.NewPoint(0, 0), float64(defaultDims.Width), float64(defaultDims.Height))
+		s := shape.NewShape(shapeType, contentBox)
 
+		paddingX, paddingY := s.GetDefaultPadding()
+		// give shapes with icons extra padding to fit their label
+		if obj.Attributes.Icon != nil {
+			labelHeight := float64(obj.LabelDimensions.Height)
+			// Evenly pad enough to fit label above icon
+			paddingX += labelHeight
+			paddingY += labelHeight
+		}
 		switch shapeType {
-		case d2target.ShapeSquare, d2target.ShapeCircle:
-			if desiredWidth != 0 || desiredHeight != 0 {
-				paddingX = 0.
-				paddingY = 0.
+		case shape.TABLE_TYPE, shape.CLASS_TYPE, shape.CODE_TYPE, shape.IMAGE_TYPE:
+		default:
+			if obj.Attributes.Link != "" {
+				paddingX += 32
+			}
+			if obj.Attributes.Tooltip != "" {
+				paddingX += 32
 			}
+		}
 
-			sideLength := math.Max(obj.Width+paddingX, obj.Height+paddingY)
+		// width and height will be set to desired values unless it is too small to fit the content with 0 padding
+		// in which case that minimum size will be used
+		if desiredWidth != 0 || desiredHeight != 0 {
+			paddingX = 0.
+			paddingY = 0.
+			if dslShape != d2target.ShapeText && obj.Attributes.Label.Value != "" {
+				contentBox.Width -= float64(INNER_LABEL_PADDING)
+				contentBox.Height -= float64(INNER_LABEL_PADDING)
+			}
```

Comment:
i think u should just wrap the place where inner_label_padding was added in an if statement

https://github.com/gavin-ts/d2/blob/d841a3423fea26590c444269bf8c3b983b395501/d2graph/d2graph.go#L1212

instead of undoing it here

Reply:
hmm okay I can try refactoring it but it is more complex than just adding an if because `getDefaultSize` has different behaviors for each shape and current behavior assumes it is called with the label padding.

Reply:
i mean if it adds complexity, then we shouldn't do it